File: android/app/src/main/java/com/scamshield/app/detection/model/DetectionModels.kt
package com.scamshield.app.detection.model

import com.google.gson.annotations.SerializedName

/**
 * Data models for the on-device detection pipeline.
 * Mirrors the backend Pydantic contracts for consistency.
 *
 * Gson emits Kotlin property names by default; FastAPI expects snake_case keys — use [@SerializedName].
 */

/** Source of the incoming message */
enum class InputSource(val value: String) {
    SMS("sms"),
    EMAIL("email"),
    NOTIFICATION("notification"),
    MANUAL("manual")
}

/** A single matched rule from the detection engine */
data class MatchedRule(
    val rule: String,
    val weight: Float,
    val category: String? = null
)

/** Output of the on-device RuleEngine */
data class RuleVerdict(
    val isSuspicious: Boolean = false,
    val confidence: Float = 0f,
    val matchedRules: List<MatchedRule> = emptyList(),
    val categoryHints: List<String> = emptyList()
)

/** API request body for /api/v1/detect */
data class DetectionRequest(
    val text: String,
    val source: String = "manual",
    val sender: String? = null,
    @SerializedName("language_hint") val languageHint: String? = null,
    @SerializedName("rule_verdict") val ruleVerdict: RuleVerdictDto? = null,
    @SerializedName("privacy_mode") val privacyMode: Boolean = false,
    val media_type: String? = null,
    val media_filename: String? = null
)

data class FeedbackRequest(
    val message_id: String,
    val label: String,
    val notes: String = ""
)

/** Rule verdict DTO for API serialization */
data class RuleVerdictDto(
    val is_suspicious: Boolean,
    val confidence: Float,
    val matched_rules: List<MatchedRuleDto>,
    val category_hints: List<String>
)

data class MatchedRuleDto(
    val rule: String,
    val weight: Float,
    val category: String? = null
)

/** Full detection response from the backend */
data class DetectionResponse(
    val verdict: DetectionVerdictDto,
    val processing_time_ms: Float = 0f,
    val llm_pending: Boolean = false
)

data class DetectionVerdictDto(
    val message_id: String = "",
    val is_scam: Boolean = false,
    val confidence: Float = 0f,
    val category: String = "unknown",
    val risk_level: String = "safe",
    val reasoning: String = "",
    val explanation: String = "",
    val recommended_action: String = "none",
    val language: String = "en",
    val detection_mode: String = "rule_only",
    val red_flags: List<String> = emptyList(),
    val sender_score: Float = 0.5f,
    val link_risk_score: Float = 0f,
    val domain_flags: List<String> = emptyList(),
    val media_risk_score: Float = 0f,
    val media_flags: List<String> = emptyList(),
    val media_type_detected: String? = null,
    val explainability: ExplainabilityDto = ExplainabilityDto()
)

data class ExplainabilityDto(
    val rule_score: Float = 0f,
    val llm_score: Float = 0f,
    val sender_reputation: Float = 0.5f,
    val final_score: Float = 0f,
    val reasons: List<String> = emptyList()
)

/** Converts local RuleVerdict to API-compatible DTO */
fun RuleVerdict.toDto(): RuleVerdictDto = RuleVerdictDto(
    is_suspicious = isSuspicious,
    confidence = confidence,
    matched_rules = matchedRules.map { MatchedRuleDto(it.rule, it.weight, it.category) },
    category_hints = categoryHints
)

File: backend/app/models/detection_models.py
"""
Detection Models — Pydantic v2 data contracts for DetectionAgent pipeline.

All input/output contracts between detection modules are defined here.
Ensures type safety, validation, and serialization consistency.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, field_validator


# --- Enums ---

class InputSource(str, Enum):
    """Source of the incoming message."""
    SMS = "sms"
    EMAIL = "email"
    NOTIFICATION = "notification"
    MANUAL = "manual"


class RiskLevel(str, Enum):
    """Risk classification tiers."""
    SAFE = "safe"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class RecommendedAction(str, Enum):
    """Recommended system action based on risk."""
    NONE = "none"
    LOG_ONLY = "log_only"
    WARN_USER = "warn_user"
    BLOCK = "block"
    BAIT = "bait"


class DetectionMode(str, Enum):
    """Which detection pipeline produced the verdict."""
    RULE_ONLY = "rule_only"
    LLM_ONLY = "llm_only"
    HYBRID = "hybrid"


# --- Input Models ---

class RawMessageInput(BaseModel):
    """Raw message received from Android client or API caller."""
    text: str = Field(..., min_length=1, max_length=5000)
    source: InputSource = InputSource.MANUAL
    sender: str | None = None
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("text")
    @classmethod
    def text_not_empty(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("Message text cannot be empty or whitespace-only")
        return v


class NormalizedInput(BaseModel):
    """Output of InputNormalizer — cleaned and enriched text."""
    message_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    text: str  # cleaned text
    original_text: str  # preserved for display
    source: InputSource
    sender: str | None = None
    language: str = "en"  # detected: "en", "hi", "hinglish"
    urls: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    normalization_applied: list[str] = Field(default_factory=list)


# --- Rule Engine Models ---

class MatchedRule(BaseModel):
    """A single rule that matched during rule-based detection."""
    rule: str  # e.g., "keyword:urgent", "pattern:shortened_url"
    weight: float = Field(ge=0.0, le=1.0)
    category: str | None = None


class RuleVerdict(BaseModel):
    """Output of the on-device RuleEngine."""
    is_suspicious: bool = False
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    matched_rules: list[MatchedRule] = Field(default_factory=list)
    category_hints: list[str] = Field(default_factory=list)


# --- LLM Models ---

class LLMVerdict(BaseModel):
    """Output of the LLM classifier."""
    is_scam: bool = False
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    category: str = "unknown"
    reasoning: str = ""
    explanation: str = ""  # user-facing
    features: dict[str, float] = Field(default_factory=dict)
    red_flags: list[str] = Field(default_factory=list)


class ExplainabilityBreakdown(BaseModel):
    rule_score: float = Field(default=0.0, ge=0.0, le=1.0)
    llm_score: float = Field(default=0.0, ge=0.0, le=1.0)
    sender_reputation: float = Field(default=0.5, ge=0.0, le=1.0)
    final_score: float = Field(default=0.0, ge=0.0, le=1.0)
    reasons: list[str] = Field(default_factory=list)


# --- Final Verdict ---

class DetectionVerdict(BaseModel):
    """Final merged detection result — returned to client."""
    message_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    is_scam: bool = False
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    category: str = "unknown"
    risk_level: RiskLevel = RiskLevel.SAFE
    reasoning: str = ""
    explanation: str = ""
    matched_rules: list[MatchedRule] = Field(default_factory=list)
    features: dict[str, float] = Field(default_factory=dict)
    red_flags: list[str] = Field(default_factory=list)
    sender_score: float = Field(default=0.5, ge=0.0, le=1.0)
    link_risk_score: float = Field(default=0.0, ge=0.0, le=1.0)
    domain_flags: list[str] = Field(default_factory=list)
    explainability: ExplainabilityBreakdown = Field(default_factory=ExplainabilityBreakdown)
    recommended_action: RecommendedAction = RecommendedAction.NONE
    language: str = "en"
    source: InputSource = InputSource.MANUAL
    detection_mode: DetectionMode = DetectionMode.RULE_ONLY

    @property
    def is_actionable(self) -> bool:
        """Whether this verdict should trigger a user-facing action."""
        return self.risk_level in (RiskLevel.MEDIUM, RiskLevel.HIGH)


# --- API Request/Response ---

class DetectionRequest(BaseModel):
    """API request body for /api/v1/detect endpoint."""
    text: str = Field(..., min_length=1, max_length=5000)
    source: InputSource = InputSource.MANUAL
    sender: str | None = None
    language_hint: str | None = None
    rule_verdict: RuleVerdict | None = None
    privacy_mode: bool = False


class DetectionResponse(BaseModel):
    """API response from /api/v1/detect endpoint."""
    verdict: DetectionVerdict
    processing_time_ms: float = 0.0
    llm_pending: bool = False  # True if LLM result will arrive async


class FeedbackRequest(BaseModel):
    message_id: str
    label: str = Field(..., pattern="^(safe|scam)$")
    notes: str = ""
